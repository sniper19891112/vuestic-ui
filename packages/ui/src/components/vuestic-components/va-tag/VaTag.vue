<template>
  <div
    v-if="valueComputed"
    class="va-tag"
    :class="computedClass"
    :style="computedStyle"
  >
    <div class="va-tag__content">
      <va-icon
        v-if="c_icon"
        class="va-tag__icon"
        :name="icon"
        :size="iconSize"
      />
      <slot></slot>
      <va-icon
        v-if="c_closeable"
        class="va-tag__close-icon"
        @click="close()"
        name="close"
        size="20px"
      />
    </div>
  </div>
</template>

<script lang="ts">
import VaIcon from '../va-icon/VaIcon.vue'
import { ColorThemeMixin } from '../../../services/ColorThemePlugin'
import { StatefulMixin } from '../../vuestic-mixins/StatefullMixin/StatefulMixin'
import { makeContextablePropsMixin } from '../../context-test/context-provide/ContextPlugin'
import { Component, Mixins } from 'vue-property-decorator'

const TagPropsMixin = makeContextablePropsMixin({
  value: { type: Boolean, default: true },
  closeable: { type: Boolean, default: false },
  color: { type: String, default: '' },
  outline: { type: Boolean, default: false },
  flat: { type: Boolean, default: false },
  icon: { type: String, default: '' },
  iconSize: { type: [Number, String], default: '18px' },
  disabled: { type: Boolean, default: false },
  size: {
    type: String,
    default: 'medium',
    validator: (value: string) => {
      return ['medium', 'small', 'large'].includes(value)
    },
  },
})

@Component({
  name: 'VaTag',
  components: { VaIcon },
})
export default class VaTag extends Mixins(
  StatefulMixin,
  ColorThemeMixin,
  TagPropsMixin,
) {
  get computedClass () {
    return {
      'va-tag--small': this.c_size === 'small',
      'va-tag--large': this.c_size === 'large',
      'va-tag--disabled': this.c_disabled,
    }
  }

  get computedStyle () {
    return {
      backgroundColor: this.c_flat
        ? ''
        : !this.c_outline
          ? this.colorComputed
          : '',
      borderColor: !this.c_flat
        ? this.colorComputed
        : '',
      color: this.c_outline || this.c_flat
        ? this.colorComputed
        : '#fff',
    }
  }

  close () {
    if (!this.c_disabled) {
      this.valueComputed = false
    }
  }
}
</script>

<style lang="scss">
@import "../../vuestic-sass/resources/resources";

.va-tag {
  display: inline-flex;
  border: 0.125rem solid transparent;
  border-radius: 2rem;
  width: auto;
  height: auto;
  min-width: initial;
  min-height: initial;
  margin: 0;
  padding: 0 0.5rem;

  &__content {
    display: flex;
    align-items: center;
    padding: auto;
    line-height: 1.6;
  }

  &__icon {
    padding-right: 0.375rem;
  }

  &__close-icon {
    cursor: pointer;
    margin-left: 0.375rem;
  }

  &--small {
    height: 1.5rem;
  }

  &--large {
    height: 2.5rem;
  }

  &--disabled {
    @include va-disabled;
  }
}
</style>
