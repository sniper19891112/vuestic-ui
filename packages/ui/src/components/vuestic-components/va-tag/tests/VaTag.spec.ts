import { mount, shallowMount } from '@vue/test-utils'
import VaTag from '../VaTag.vue'

describe('VaTag', () => {
  it('should render without an error', () => {
    const wrapper = mount(VaTag)
    expect(wrapper.isVueInstance()).toBeTruthy()
  })
  /* 2. Button sizes */

  it('large tag', () => {
    const wrapper = shallowMount(VaTag, {
      propsData: { size: 'large' },
    })
    expect(wrapper.find(VaTag).classes()).toContain('va-tag--large')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('small tag', () => {
    const wrapper = shallowMount(VaTag, {
      propsData: { size: 'small' },
    })
    expect(wrapper.find(VaTag).classes()).toContain('va-tag--small')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should be a link', () => {
    const wrapper = shallowMount(VaTag, {
      propsData: {
        href: 'http://epic-spinners.epicmax.co/',
      },
    })
    expect(wrapper.is('a')).toBe(true)
  })
})
