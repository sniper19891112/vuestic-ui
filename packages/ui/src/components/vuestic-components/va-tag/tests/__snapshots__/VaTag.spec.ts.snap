// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VaTag large tag 1`] = `
<div tabindex="0" class="va-tag va-tag--large" style="color: rgb(255, 255, 255); background: rgb(0, 0, 0); box-shadow: 0 0.125rem 0.19rem 0 rgba(0,0,0,0.4);">
  <!---->
  <div class="va-tag__content"></div>
  <!---->
</div>
`;

exports[`VaTag small tag 1`] = `
<div tabindex="0" class="va-tag va-tag--small" style="color: rgb(255, 255, 255); background: rgb(0, 0, 0); box-shadow: 0 0.125rem 0.19rem 0 rgba(0,0,0,0.4);">
  <!---->
  <div class="va-tag__content"></div>
  <!---->
</div>
`;
